### ISSUE TITLE
Result of query is mixed array

### ISSUE THREAD

**Issue by slaviczavik** [porsager/postgres#147]
According to docs:

```js
// other.js
const sql = require('./db.js')

const users = await sql`
  select name, age from users
`
// users: [{ name: 'Murray', age: 68 }, { name: 'Walter', age: 78 }]
```

the result of a sql query call should be an array of rows, but in my case (v1.0.2) this array is returned:

```js
[
  { name: 'Murray', age: 68 },
  { name: 'Walter', age: 78 },
  count: 2,
  command: 'SELECT'
]
```

This is really unfortunate because I can't easily return the data without count and command attributes.

**Comment by Minigugus** [comment]
If you plan to serialize this array in JSON, theses properties are not an issue:
```js
// other.js
const sql = require('./db.js')

const users = await sql`
  select name, age from users
`;
console.info(JSON.stringify(user));
// prints "[{"name":"Murray","age":68},{"name":"Walter","age":78}]"
```

you can also clone the array: `[...users]`

**Comment by slaviczavik** [comment]
Sorry, but this is a terrible way how to return data. Something like the code below would be better.
```js
{
  rows: [{ name: 'Murray', age: 68 }, { name: 'Walter', age: 78 }]
  count: 2,
  command: 'SELECT'
}
```

**Comment by porsager** [comment]
Could you try to describe your use case? 

It would also be great if you could describe in a constructive way why you find it terrible?

**Comment by slaviczavik** [comment]
Stringify result or destructing the array for getting the result is an unnecessary step. Second, the array with count and command attribute just doesn't looks like a valid array.

**Comment by porsager** [comment]
I'm still very curious what you need it for? 

Is it because you don't like the looks of it when you console.log it?

**Comment by Minigugus** [comment]
> Second, the array with count and command attribute just doesn't looks like a valid array.

It doesn't looks valid but it is. It's a part of the JavaScript design. It also works with functions.

If it's just the console.log output you dislike, maybe it is possible to set `enumerable: false` on `count` and `command` properties, and even add a `rows` property to the array that return a clean array?

**Comment by Olena-Stotska** [comment]
I'll try to explain this, in a constructive way:

The point is that if you need an object, you need to use an object and not an array. Would you use `Set` when you need `HashMap` add implement additional functionality on top `Set`? I don't think so.

**TL;DR**:
There are a bunch of issues with this decision:
1. It seems that `count` duplicates native `Array.prototype.length`. If so, then why?
2. `command` and `count` are implementation details of postgres.js . So, why do you expose this if library consumers don't need it? If for some reason, you need them internally why not to expose this through an object? 
3. We can't safely return the return result from postgres.js further without a proper array cleanup because we would break encapsulation. So, SQL query request, we need to clone this array. To clone an array, we spend additional O(n)
4. Very likely V8 can't optimize this array because it has additional properties (read [this for details](https://v8.dev/blog/fast-properties))

All this can be fixed by replacing an array with an object that contains a clean array inside. **Any reasons** to not do this?

**Comment by Minigugus** [comment]
> 1. It seems that count duplicates native Array.prototype.length. If so, then why?

Not really, it returns the number of affecter rows for `UPDATE` and `INSERT` requests. That's also the reason why not returning a simple array is complicated, especially with the tagged template pattern (the SQL request type cannot be determined statically).

> 2. command and count are implementation details of postgres.js . So, why do you expose this if library consumers don't need it? If for some reason, you need them internally why not to expose this through an object?

Like said above, `count` is required in some case for users that wants to check that an `UPDATE` for instance really updated the expected number of rows, and returning an empty array in this case would be a bit stupid. Similary, the `command` property is useful for logging purposes.

Then maybe it should be an object but the tagged template function always returned an array; change it now is a real breaking change. This option should be considered carefully.

> 3. We (library consumers) can't safely return the return result from postgres.js further without a proper array cleanup because we would break encapsulation. So, on every SQL query request, we need to clone this array. To clone an array, we spend additional O(n)

I think a lot of users use `postgres` directly in their applications, then encapsulation doesn't care as much as for a library. In my opinion, `postgres` targets mainly applications, not libraries (for instance it's hard to use it for proper dynamic queries building). I think what is this design goal should be discussed.

> 4. Very likely V8 can't optimize this array because it has additional properties (not an expert, read this for details :))

I can't argue with that :sweat_smile:


To conclude, I understand your point, however the fact that the tagged template function of `postgres` returns an array is a very sensitive subject, not trivial to solve. In my mind, a simple `rows` property is not satisfing either, as it would require a lot of users to add `.rows` to their existing and new code. I think the solution is between the augmented array and a plain object (an iterable object, or an ES6 Proxy for instance :wink:)

*It's mainly an opinion-based discussion I think, feel free to share yours, maybe it can help* :wink:

**Comment by porsager** [comment]
Thank you for the deeper explanation @Olena-Stotska !

> The point is that if you need an object, you need to use an object and not an array. Would you use Set when you need HashMap and implement additional functionality on top of Set? I don't think so.

Just to be clear - This is an array. Also - an array in javascript **is** an object. The fact that this array has 2 extra properties does not make it less of an array. 

1. `count` is the number of results from Postgres, and highly useful in application scenarios. For instance if you have `insert` / `update` queries you only get the count of affected rows back, but not any actual rows, unless you do `returning ...` in your query.

3. I'm still very curious where it is you want to "safely return it" to that becomes a problem? As @Minigugus said, `toJSON` gives you the array.
 
4. This doesn't have any negative effect on performance.

I'm sorry, but I still don't see there is anything to "fix" here. 

The first reason not to change this is not to fix what isn't broken 😉
The second reason is that this is a succinct result signature that works for all queries. You can easily destructure the result for the common scenarios like accessing the first element of the array, or the count if you need to check how many rows where updated. `([first]) => ...` or `({ count }) => ...`

**Comment by Minigugus** [comment]
@porsager Destructuring arrays in JS works with any `Iterable` object, and destructuring objects with any object, therefore `([first]) => ...` and `({ count }) => ...` should work for any iterable object with a `count` count property, an augmented array is not needed :neutral_face: 

The only issue is for random access :disappointed:, so as of now, the best solution seems to be the array. :wink:

**Comment by porsager** [comment]
> The only issue is for random access 😞

Not sure I understand - do you mean accessing a specific position in the array other than the first?

**Comment by Minigugus** [comment]
> Not sure I understand - do you mean accessing a specific position in the array other than the first?

Yes, using a normal object. Only the array allows fast initialization and access in this case (I think). :wink:

**Comment by porsager** [comment]
Ah, well that's actually doable :P Or is this not what you mean?
```js
const { 2: x } = ['a', 'b', 'c'] 
// x = 'c'
```

or actually random, but that's just silly :P
```js
const xs = ['a', 'b', 'c', 'd']
    , { [Math.floor(Math.random() * xs.length)]: x } = xs
// x = a | b | c | d ?
```

**Comment by Minigugus** [comment]
I'm sorry, not in this way x) I'm sorry my english is not that good, my bad :sweat_smile: 

I meant the following code is slower than if `obj` where an array:
```js
const obj = {}; // can be faster with `= [];` instead
for (let i = 0; i < 99999999; i++)
  obj[î] = i;

console.log(obj[Math.floor(Math.random() * 99999999)]);
```

If we can solve this performance issue then we can drop the array and use an augmented class instead (iterable, with `toJSON`, and so on) :wink:

**Comment by uasan** [comment]
Will this option suit everyone?
```JavaScript
class Result extends Array {
   count = 0;
   command = '';
}
```

**Comment by porsager** [comment]
Hehe, sorry I'm still not getting it. What are we trying to solve? And is there an actual performance issue?

**Comment by uasan** [comment]
I think the reason for the indignation, in the absence of an explicit type of object, the Result type solves this problem and this is an Array )

**Comment by porsager** [comment]
I'm still not convinced how that is a problem. It seems like it's a matter of taste, and to me a class almost always leaves a bad taste in my mouth.

Also, just did some initial tests with your Result class suggestion, which shows it to be ~12% slower than an array.

**Comment by slaviczavik** [comment]
I'm just curious about how the array with the object attributes (count and command) is constructed. Can you show me an example?

**Comment by porsager** [comment]
Sure thing :) 
```
result = []
```

**Comment by slaviczavik** [comment]
How you put the count attribute there? I'd like to see, how you construct array like this [{row}, count: 1, command: 'SELECT'].

**Comment by porsager** [comment]
Ah, right. They're set when they arrive from the database.

Basically [like this:](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgCsEAaEAYwHs0AXGWvKtOGgAgCd4BXKNgXlbIAugB00nODxr4qXWqwEBGAAxiJUmZSxYMaACYLWAcgCqABQAiAQQAqAUSNixABwAUY1h268SrAPR+gqJonuq8rBDMNLrkMJRgrFbs7BgAnr4BrDTsXDAeXpK8mnI0GYEq+WHSVNq6BpmmlrYOYgCUIGRwMLDkNBDUCIggAJyIiiAAvkITQA)
```js
const result = []
result.count = ...
result.command = ...
```

**Comment by slaviczavik** [comment]
Ok, but there is one small issue. What if in some point in future will be change in Array object and count or command property won't be "available", basically it would be same as overriding a length attribute for example. And that doesn't looks right to me.

**Comment by porsager** [comment]
Yes, that's a valid point, but not quite the same as smooshgate. We're not extending prototypes here, and we're not in the browser.

I highly doubt they would ever make them non writable or configurable, and if that would be the case it's not hard to fix if that should ever happen. That's a tradeoff I'm fine with.

**Comment by Minigugus** [comment]
To get everyone to agree (or not), maybe we could use a Proxy :laughing: 
```js
const array = [42]
const result = new Proxy({ count: 42, command: 'SELECT 42' }, {
  get(target, prop, receiver) {
    if (prop in array)
      return Reflect.get(array, prop, array)
    return Reflect.get(target, prop, receiver)
  }
})

const [first1] = result; // works
const first2 = result[0]; // also works
```

However, I'm just a little sceptical about the performance...

**Comment by stalniy** [comment]
Ok, I see one more issue :) `count` is a misleading name. It should be `affectedRowsCount` or something similar. In that way, it will make more sense for people and then we will not be confused because of the naming.

@Minigugus regarding returning array vs object. There is always a way to implement this with meaningful defaults (i.e., configuration). In majority of cases (at least for select), we need an array so by default `sql` should return an array. If case we need affectedRowsCount there should be a setting or option to do this, so it's explicit and clear.

My proposal is this:

1. Return always an array from `sql` tag
2. Provide a helper that returns affected rows: 
    ```js
    await sql.detailedResult`
      UPDATE ....
    `
    ```

The min change which I think should be done is to rename `count` to `affectedRowsCount`.

**Please** DO NOT use `Proxy` :)

**Comment by porsager** [comment]
I feel this is turning into bikeshedding at it's best. Could someone highlight to me where this has been confusing / a problem except for the fact that a console.log looks like this: `[...rows, { count: 10, command: 'SELECT }]`?

**Comment by uasan** [comment]
> Also, just did some initial tests with your Result class suggestion, which shows it to be ~12% slower than an array.

It's weird but I have a better idea )

I suggest using `WeakMap` where the key is an array, the value is an object with meta information from the result, and add  helpers functions.

```JavaScript
//lib level
const meta = new WeakMap();

export const getMeta = result => meta.get(result);
```

```JavaScript
//user level
import { getMeta } from 'postgres';

getMeta(await sql`UPDATE ....`);
```

**Comment by stalniy** [comment]
Let's make that properties to be non-enumerable and define them with help of `Object.defineProperty`. Doing so, nobody who doesn't need/know about these properties will not see them even when he tries to do `console.log(rows)`. Also, the shape of return value will be preserved:

```js
const rows = Object.defineProperties([], {
  count: { value: 2 },
  command: { value: "SELECT" }
});

console.log(rows);
```

@porsager any thoughts?

**Comment by porsager** [comment]
@stalniy I like that! So obvious in hindsight 🙂

**Comment by dirk-ecker** [comment]
Sorry to comment an issue that is closed, but it the result can't be used in a for of loop and that is really bad.

`for (const siteId of siteIds) {`

**Comment by porsager** [comment]
No worries, but that's not right, it'll work fine in a `for of` loop. Are you sure you're not simply forgetting an `await` and therefore trying to use the promise?

**Comment by dirk-ecker** [comment]
> No worries, but that's not right, it'll work fine in a `for of` loop. Are you sure you're not simply forgetting an `await` and therefore trying to loop the promise?

Oh, yes sorry, that was my problem. I saw it some minutes after commenting.

**Comment by porsager** [comment]
Ok, great you got it working 👍

**Comment by dirk-ecker** [comment]
Yes only to blind to see, that the await is missing, as you expected.
